edx_arch_experiments/scripts/get_action_errors.py

```diff
@@ -1,138 +1,262 @@
+        """
+        backoff_s = None  # exponential backoff in seconds, or None if not in effect
+        while True:
+            # If GitHub has told us how long to wait, use that instead of the
+            # current exponential backoff value.
+            if self.github_sleep_until_s is not None:
+                # Add slop to prevent tight loop on expiry
+                sleep_s = self.github_sleep_until_s + 5.0 - time.time()
+            else:
+                sleep_s = backoff_s
+
+            if sleep_s and sleep_s > 0:
+                time.sleep(sleep_s)
+
+            response = requests.get(url, params=params, headers=self.api_headers, timeout=20.0)
+
+            # Update rate-limiting data for next call
+            if out_of_requests := (response.headers.get('x-ratelimit-remaining') == '0'):
+                self.github_sleep_until_s = int(response.headers.get('x-ratelimit-reset'))
+                print(
+                    "Reached rate limit. "
+                    f"Will wait {int(self.github_sleep_until_s - time.time())} seconds "
+                    "before next request."
+                )
+            else:
+                self.github_sleep_until_s = None
+                backoff_s = None
+
+            if response.status_code == 200:
+                # We're good to go!
+                return response
+            elif out_of_requests:
+                # We got an error and have been informed we're out of requests.
+                # (Should be a 429 or 403 according to GitHub's docs.) We'll try
+                # again.
+                continue
+            elif response.status_code == 429:
+                # It's possible that GitHub might give us a 429 without the
+                # expected rate-limiting headers.
+                print("Rate-limited without timing hint; performing exponential backoff.")
+                backoff_s = 2 * backoff_s if backoff_s else 4
+                continue
+            else:
+                # Generic error case
+                response.raise_for_status()
+
+    def _write_json(self, data, *path_parts):
+        """
+        Write data as pretty-printed JSON to the given path (joining as needed).
+        """
+        with open(path.join(*path_parts), 'w') as f:
+            json.dump(data, f, sort_keys=True, indent=2)
+
+    def _log_attempt(self, attempt, workflow_dir):
+        """
+        Log this attempt and all of its checks and annotations.
+        """
+        attempt_dir = _ensure_dir(workflow_dir, f"attempt_{attempt['run_attempt']}")
+
+        attempt_file = path.join(attempt_dir, f"attempt.json")
+        if path.isfile(attempt_file):
+            print("Attempt already fully downloaded; skipping.")
+            return
+
+        # Get the checks associated with this workflow run -- this
+        # includes output title, summary, and text.
+        for check_run in self._github_get(attempt['check_suite_url'] + '/check-runs').json()['check_runs']:
```

Comment:
Author's note: I'm not sure this is actually working properly -- different attempts on the same workflow are getting the same check runs.

Reply:
Asking about this here: https://github.com/orgs/community/discussions/103026